### PR TITLE
Add 2 factor authentication using OTP

### PR DESCRIPTION
**Pull request by hughsie**
Fixes https://github.com/hughsie/lvfs-website/issues/239

### CODE REVIEW

File: app/templates/profile.html [hunk 1]
```diff
@@ -36,11 +36,66 @@ <h1>Modify Local User</h1>
...
+    <div class="modal-content">
+      <div class="modal-header">
+        <h5 class="modal-title" id="otpSetupModalLabel">Setup OTP</h5>
+        <button type="button" class="close" data-dismiss="modal" aria-label="Close">
+          <span aria-hidden="true">&times;</span>
+        </button>
+      </div>
+      <div class="modal-body">
+        <p>
+          Please start FreeOTP and scan the following QR Code:
```
Comment (superm1) [inline app/templates/profile.html]: I suspect most people are going to use something like Google Authenticator or something else they already have for other sites maybe it's worth making it clear that they can use any client that supported this.

Comment (hughsie) [reply app/templates/profile.html]: `Please start FreeOTP, Google Authenticator or any other suitable application and scan the following QR Code` ? I could hyperlink FreeOTP and Google Authenticator too.

Comment (superm1) [reply app/templates/profile.html]: Yeah I think that's good

**Comment by superm1** [comment]
I feel like this kind of attribute is actually something the manager should be able to "mandate" but then that means a new state in state machine for enabled but not working.

**Comment by hughsie** [comment]
> this kind of attribute is actually something the manager should be able to "mandate" 

This gets really hard, as a manager adding accounts for users would have to get the user to scan the QR code *before* the initial login. Or I suppose we could deal with the 2FA setup on first login, although that needs some careful design to avoid it being insecure. I think it's a new PR for new functionality, not this one IMHO.

**Comment by superm1** [comment]
Yeah that's what I was thinking too.  It's definitely a new state to define on doing it for first login, similar to a mandatory password change request.  I don't know that it's necessary, just wanted to make sure it's discussed.

File: app/templates/profile.html (hunk 1 above)
Comment (hughsie) [reply app/templates/profile.html]: Fixed.

**Comment by hughsie** [comment]
Other than the organisation-policy for 2FA, no other objections? I'm not 100% sure about the login box now:

screenshot from 2019-01-22 15-22-50

I'm quite keen on keeking things simple and not do client-side JS to hide/show the OTP if it's not configured. Hopefully longer term more local accounts than not will have 2FA enabled, or it'll be a constant reminder to do so.

**Comment by hughsie** [comment]
> 2FA should still work with OAuth right?

2FA and OAuth are mutually exclusive from the LVFS point of view. If you want to use 2FA using OAuth you need to configure it Azure-side so-to-speak.

> What about making it a 2 stage login screen

I thought a lot about this; I didn't find any way it would be secure. I'd have to store a login cookie that indicated that the user was "half" logged on, and that broke quite a few assumptions in the flask code that I didn't quite understand.

**Comment by superm1** [comment]
> 2FA and OAuth are mutually exclusive from the LVFS point of view. If you want to use 2FA using OAuth you need to configure it Azure-side so-to-speak.

Right, that's what I was thinking. So it needs to be pretty explicit which fields go to which login.

> I thought a lot about this; I didn't find any way it would be secure. I'd have to store a login cookie that indicated that the user was "half" logged on, and that broke quite a few assumptions in the flask code that I didn't quite understand.

Alright so what about re-organizing the page so that it's a "left and right" split with a top delimitter.

```
Header text
------------
      |
OAuth | Password login
      |
```

Then in the password login side of the window you make the 2FA thing indented and a different color.
The indentation will look better with the horizontal view.

**Comment by hughsie** [comment]
> Alright so what about re-organizing the page

Maybe this is the straw that breaks the camels back. What about:

* Making the `Login` button turn into a dropdown, with two sub-entries, `Azure AD` and `Local Account`, possibly with the local option first.
* The `Azure AD` button would take you to the SSO if applicable
* The `Local Account` would show the username/password/otp, perhaps remaining as a modal dialog.

**Comment by superm1** [comment]
I can tell you that's _likely_ to prompt anyone who does use Azure to start trying to enter domain credentials into LVFS.  Until azure was made more prominent I recall that we were having that happen.

**Counter-proposal:**
Make the login page two stage.
1. user enters email address
2. LVFS either redirects to oauth provider or returns back a page to do login process.
3. On returned login process page LVFS can return 2FA section in the page if it's needed.

I would think that should avoid breaking security (needing to track a cookie or anything like that).

**Comment by hughsie** [comment]
> for your ODMs?

Ignore me, they're probably not using OAuth.

**Comment by hughsie** [comment]
@superm1  -- something like this?

screenshot from 2019-01-23 11-58-31
screenshot from 2019-01-23 11-58-22

**Comment by hughsie** [comment]
Okay, comments welcome ^^

**Comment by superm1** [comment]
That LGTM

**Pull request merged**